### PR TITLE
Jit64: Optimize divwx

### PR DESCRIPTION
**Pull request by Sintendo**
Optimize various occurrences of signed 32-bit integer division. Most notably, it implements the multiplication by magic number method to optimize divisions with a constant divisor (the Zelda games like to do this a bunch).

---

**Constant dividend**

<details><summary>Dividend != 0x80000000</summary>

Before:
```
B8 54 D3 E6 02       mov         eax,2E6D354h
85 FF                test        edi,edi
74 0C                je          overflow
3D 00 00 00 80       cmp         eax,80000000h
75 0C                jne         normal_path
83 FF FF             cmp         edi,0FFFFFFFFh
75 07                jne         normal_path
overflow:
C1 F8 1F             sar         eax,1Fh
8B F8                mov         edi,eax
EB 05                jmp         done
normal_path:
99                   cdq
F7 FF                idiv        eax,edi
8B F8                mov         edi,eax
done:
```

After:
```
85 FF                test        edi,edi
75 04                jne         normal_path
33 FF                xor         edi,edi
EB 0A                jmp         done
normal_path:
B8 54 D3 E6 02       mov         eax,2E6D354h
99                   cdq
F7 FF                idiv        eax,edi
8B F8                mov         edi,eax
done:
```
</details>

<details><summary>Dividend == 0</summary>

Before:
```
B8 00 00 00 00       mov         eax,0
85 F6                test        esi,esi
74 0C                je          overflow
3D 00 00 00 80       cmp         eax,80000000h
75 0C                jne         normal_path
83 FE FF             cmp         esi,0FFFFFFFFh
75 07                jne         normal_path
overflow:
C1 F8 1F             sar         eax,1Fh
8B F8                mov         edi,eax
EB 05                jmp         done
normal_path:
99                   cdq
F7 FE                idiv        eax,esi
8B F8                mov         edi,eax
done:
```

After:
Nothing!
</details>

---

**Constant divisor**

<details><summary>Division by 1</summary>

Before:
```
41 BF 01 00 00 00    mov         r15d,1
41 8B C5             mov         eax,r13d
45 85 FF             test        r15d,r15d
74 0D                je          overflow
3D 00 00 00 80       cmp         eax,80000000h
75 0E                jne         normal_path
41 83 FF FF          cmp         r15d,0FFFFFFFFh
75 08                jne         normal_path
overflow:
C1 F8 1F             sar         eax,1Fh
44 8B F8             mov         r15d,eax
EB 07                jmp         done
normal_path:
99                   cdq
41 F7 FF             idiv        eax,r15d
44 8B F8             mov         r15d,eax
done:
```

After:
```
45 8B FD             mov         r15d,r13d
```
</details>

<details><summary>Division by 2</summary>

Before:
```
41 BE 02 00 00 00    mov         r14d,2
41 8B C2             mov         eax,r10d
45 85 F6             test        r14d,r14d
74 0D                je          overflow
3D 00 00 00 80       cmp         eax,80000000h
75 0E                jne         normal_path
41 83 FE FF          cmp         r14d,0FFFFFFFFh
75 08                jne         normal_path
overflow:
C1 F8 1F             sar         eax,1Fh
44 8B F0             mov         r14d,eax
EB 07                jmp         done
normal_path:
99                   cdq
41 F7 FE             idiv        eax,r14d
44 8B F0             mov         r14d,eax
done:
```

After:
```
45 8B F2             mov         r14d,r10d
41 C1 EE 1F          shr         r14d,1Fh
45 03 F2             add         r14d,r10d
41 D1 FE             sar         r14d,1
```
</details>

<details><summary>Division by 4 (power of two)</summary>

Before:
```
41 BD 04 00 00 00    mov         r13d,4
41 8B C0             mov         eax,r8d
45 85 ED             test        r13d,r13d
74 0D                je          overflow
3D 00 00 00 80       cmp         eax,80000000h
75 0E                jne         normal_path
41 83 FD FF          cmp         r13d,0FFFFFFFFh
75 08                jne         normal_path
overflow:
C1 F8 1F             sar         eax,1Fh
44 8B E8             mov         r13d,eax
EB 07                jmp         done
normal_path:
99                   cdq
41 F7 FD             idiv        eax,r13d
44 8B E8             mov         r13d,eax
done:
```

After:
```
45 85 C0             test        r8d,r8d
45 8D 68 03          lea         r13d,[r8+3]
45 0F 49 E8          cmovns      r13d,r8d
41 C1 FD 02          sar         r13d,2
```
</details>

<details><summary>Division by 30307 (arbitrary constant)</summary>

Before:
```
41 BA 63 76 00 00    mov         r10d,7663h
41 8B C5             mov         eax,r13d
45 85 D2             test        r10d,r10d
74 0D                je          overflow
3D 00 00 00 80       cmp         eax,80000000h
75 0E                jne         normal_path
41 83 FA FF          cmp         r10d,0FFFFFFFFh
75 08                jne         normal_path
overflow:
C1 F8 1F             sar         eax,1Fh
44 8B C0             mov         r8d,eax
EB 07                jmp         done
normal_path:
99                   cdq
41 F7 FA             idiv        eax,r10d
44 8B C0             mov         r8d,eax
done:
```

After:
```
49 63 C5             movsxd      rax,r13d
4C 69 C0 65 6B 32 45 imul        r8,rax,45326B65h
C1 E8 1F             shr         eax,1Fh
49 C1 F8 2D          sar         r8,2Dh
44 03 C0             add         r8d,eax
```
</details>

<details><summary>Division by 30323 (arbitrary constant)</summary>

Before:
```
41 BA 73 76 00 00    mov         r10d,7673h
41 8B C5             mov         eax,r13d
45 85 D2             test        r10d,r10d
74 0D                je          overflow
3D 00 00 00 80       cmp         eax,80000000h
75 0E                jne         normal_path
41 83 FA FF          cmp         r10d,0FFFFFFFFh
75 08                jne         normal_path
overflow:
C1 F8 1F             sar         eax,1Fh
44 8B C0             mov         r8d,eax
EB 07                jmp         00000000161737E7
normal_path:
99                   cdq
41 F7 FA             idiv        eax,r10d
44 8B C0             mov         r8d,eax
done:
```

After:
```
49 63 C5             movsxd      rax,r13d
4C 69 C0 19 25 52 8A imul        r8,rax,0FFFFFFFF8A522519h
49 C1 E8 20          shr         r8,20h
44 03 C0             add         r8d,eax
C1 E8 1F             shr         eax,1Fh
41 C1 F8 0E          sar         r8d,0Eh
44 03 C0             add         r8d,eax
```
</details>

---

**Division without constants**

<details><summary>Micro-optimization</summary>

Before:
```
41 8B C7             mov         eax,r15d
45 85 C0             test        r8d,r8d
74 0D                je          overflow
3D 00 00 00 80       cmp         eax,80000000h
75 0E                jne         normal_path
41 83 F8 FF          cmp         r8d,0FFFFFFFFh
75 08                jne         normal_path
overflow:
C1 F8 1F             sar         eax,1Fh
44 8B F0             mov         r14d,eax
EB 07                jmp         done
normal_path:
99                   cdq
41 F7 F8             idiv        eax,r8d
44 8B F0             mov         r14d,eax
done:
```

After:
```
41 8B C7             mov         eax,r15d
45 85 C0             test        r8d,r8d
74 0D                je          overflow
3D 00 00 00 80       cmp         eax,80000000h
75 0B                jne         normal_path
41 83 F8 FF          cmp         r8d,0FFFFFFFFh
75 05                jne         normal_path
overflow:
C1 F8 1F             sar         eax,1Fh
EB 04                jmp         done
normal_path:
99                   cdq
41 F7 F8             idiv        eax,r8d
done:
44 8B F0             mov         r14d,eax
```
</details>

### CODE REVIEW

File: Source/Core/Core/PowerPC/JitCommon/DivUtils.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2021 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include "Common/CommonTypes.h"
+
+struct Magic
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitCommon/DivUtils.h:9]: Maybe place the contents of this file within a namespace like JitCommon, given `Magic` is quite a generic name.

File: Source/Core/Core/PowerPC/JitCommon/DivUtils.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+{
+  s32 multiplier;
+  u8 shift;
+};
+
+Magic SignedDivisionConstants(s32);
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitCommon/DivUtils.h:15]: Missing a parameter name

File: Source/Core/Core/PowerPC/JitCommon/DivUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
+// Copyright 2021 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <cstdlib>
+
+#include "Core/PowerPC/JitCommon/DivUtils.h"
+
+// Calculate the constants required to optimize a signed 32-bit integer division.
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitCommon/DivUtils.cpp:9]: This description should probably go on the prototype in the header. This way IDE builds with information on mouse hover-over also get the documentation comment.

File: Source/Core/Core/PowerPC/Jit64/Jit_Integer.cpp [hunk 4]
```diff
@@ -1340,6 +1342,208 @@ void Jit64::divwx(UGeckoInstruction inst)
...
+    }
+    else if (divisor == 1)
+    {
+      if (d != a)
+        MOV(32, Rd, Ra);
+      if (inst.OE)
+        GenerateConstantOverflow(false);
+    }
+    else if (divisor == -1)
+    {
```
Comment (MerryMage) [inline Source/Core/Core/PowerPC/Jit64/Jit_Integer.cpp:1444]: You can simplify this case to:
```
if (d != a)
  MOV(32, Rd, Ra);
NEG(32, Rd);
const FixupBranch normal = J_CC(CC_NO);
MOV(32, Rd, Imm32(0xFFFFFFFF));
if (inst.OE)
  GenerateConstantOverflow(true);
SetJumpTarget(normal);
const FixupBranch done = J();
if (inst.OE)
  GenerateConstantOverflow(false);
SetJumpTarget(done);
```

File: Source/Core/Core/PowerPC/Jit64/Jit_Integer.cpp [hunk 5]
```diff
@@ -1340,6 +1342,207 @@ void Jit64::divwx(UGeckoInstruction inst)
...
+      else
+      {
+        // Divisor is not 0, take normal path
+        normal_path = J_CC(CC_NZ);
+        // Otherwise, proceed to overflow case
+      }
+
+      // Set Rd to all ones or all zeroes
+      if (dividend & 0x80000000)
+        OR(32, Rd, Imm32(0xFFFFFFFF));
```
Comment (MerryMage) [inline Source/Core/Core/PowerPC/Jit64/Jit_Integer.cpp:1397]: I understand using `OR` for codesize here, but I'm not sure the false dependency on the previous value of Rd is worth that saving.

**Comment by MerryMage** [comment]
cmp         eax,80000000h
jne

File: Source/Core/Core/PowerPC/Jit64/Jit_Integer.cpp (hunk 5 above)
Comment (Sintendo) [reply Source/Core/Core/PowerPC/Jit64/Jit_Integer.cpp]: Yeah, you're probably right. I've replaced both instances of `OR(32, Rd, Imm32(0xFFFFFFFF));` with `MOV(32, Rd, Imm32(0xFFFFFFFF));`.

**Review by lioncash — APPROVED** [review]

**Pull request merged**